pkg/dao/rpm.go

```diff
@@ -170,3 +190,173 @@ func (r rpmDaoImpl) Search(orgID string, request api.SearchRpmRequest, limit int
+	if result.Error != nil {
+		return rowsAffected, result.Error
+	}
+
+	dbPkgs := FilteredConvert(pkgs, existingChecksums)
+
+	rowsAffected, error := r.PagedRpmInsert(&dbPkgs)
+	if error != nil {
+		return rowsAffected, result.Error
+	}
+
+	var rpmUuids []string
+	//Now fetch the uuids of all the rpms we want associated to the repository
+	result = r.db.Where("checksum in ?", checksums).Model(&models.Rpm{}).Pluck("uuid", &rpmUuids)
+	if result.Error != nil {
+		return rowsAffected, result.Error
+	}
+
+	//Delete RepositoryRpm entries we don't need
+	error = r.deleteUnneeded(repo, rpmUuids)
+	if error != nil {
+		return rowsAffected, result.Error
+	}
+
+	//Add the RepositoryRpm entries we do need
+	associations := prepRepositoryRpms(repo, rpmUuids)
+	result = r.db.Clauses(clause.OnConflict{
+		Columns:   []clause.Column{{Name: "repository_uuid"}, {Name: "rpm_uuid"}},
+		DoNothing: true}).Create(&associations)
+
+	return rowsAffected, result.Error
+}
+
+// prepRepositoryRpms  converts a list of rpm_uuids to a list of RepositoryRpm Objects
+func prepRepositoryRpms(repo models.Repository, rpm_uuids []string) []models.RepositoryRpm {
+	repoRpms := make([]models.RepositoryRpm, len(rpm_uuids))
+	for i := 0; i < len(rpm_uuids); i++ {
+		repoRpms[i].RepositoryUUID = repo.UUID
+		repoRpms[i].RpmUUID = rpm_uuids[i]
+	}
+	return repoRpms
+}
+
+// difference returns the difference between arrays a and b   (a - b)
+func difference(a, b []string) []string {
+	mb := make(map[string]struct{}, len(b))
+	for _, x := range b {
+		mb[x] = struct{}{}
+	}
+	var diff []string
+	for _, x := range a {
+		if _, found := mb[x]; !found {
+			diff = append(diff, x)
+		}
+	}
+	return diff
+}
+
+// deleteUnneeded Removes any RepositoryRpm entries that are not in the list of rpm_uuids
+func (r rpmDaoImpl) deleteUnneeded(repo models.Repository, rpm_uuids []string) error {
+	//First get uuids that are there:
+	var existing_rpm_uuids []string
+
+	r.db.Model(&models.RepositoryRpm{}).Where("repository_uuid = ?", repo.UUID).Pluck("rpm_uuid", &existing_rpm_uuids)
+
+	rpmsToDelete := difference(existing_rpm_uuids, rpm_uuids)
+
+	result := r.db.Where("repositories_rpms.repository_uuid = ? and repositories_rpms.rpm_uuid in ?", repo.UUID, rpmsToDelete).Delete(&models.RepositoryRpm{})
+	return result.Error
+}
+
+func stringInSlice(a string, list []string) bool {
+	for _, b := range list {
+		if b == a {
+			return true
+		}
+	}
+	return false
+}
+
+// FilteredConvert Given a list of yum.Package objects, it converts them to model.Rpm packages
+//	while filtering out any checksums that are in the excludedChecksums parameter
+func FilteredConvert(yumPkgs []yum.Package, excludeChecksums []string) []models.Rpm {
+	var dbPkgs []models.Rpm
+	for i := 0; i < len(yumPkgs); i++ {
+		yumPkg := yumPkgs[i]
+		if !stringInSlice(yumPkg.Checksum.Value, excludeChecksums) {
+			epoch := yumPkg.Version.Epoch
+			dbPkgs = append(dbPkgs, models.Rpm{
+				Base: models.Base{
+					UUID: uuid.NewString(),
+				},
```

Comment:
UUID is generated on creation, so assigning it here is unnecessary. Or is this UUID used somewhere before creation that I'm missing?

Reply:
thanks! I will remove it